File: server/src/routes/cs2.rs
use crate::cs2::{GSIPayload, MatchStateManager, PlayerStatsClient};
use crate::models::ApiResponse;
use rocket::serde::json::Json;
use rocket::{get, post, State};
use std::env;

#[post("/gsi", data = "<payload>")]
pub async fn receive_gsi(
    payload: Json<GSIPayload>,
    match_state: &State<MatchStateManager>,
    player_stats_client: &State<PlayerStatsClient>,
) -> Json<ApiResponse<String>> {
    // Verify auth token
    let expected_token = env::var("GSI_AUTH_TOKEN").unwrap_or_else(|_| "your_secret_token_here".to_string());

    if let Some(auth) = &payload.auth {
        if auth.token != expected_token {
            return Json(ApiResponse::error("Invalid auth token".to_string()));
        }
    } else {
        return Json(ApiResponse::error("Missing auth token".to_string()));
    }

    // Only process competitive matches
    if !payload.is_competitive() {
        return Json(ApiResponse::success("Not a competitive match, ignoring".to_string()));
    }

    // Get my Steam ID from environment
    let my_steamid = env::var("STEAM_ID").unwrap_or_default();

    // Update match state
    match_state.update_from_gsi(&payload, &my_steamid);

    // Fetch player stats for all players asynchronously
    let steamids = payload.get_all_steamids();

    tokio::spawn({
        let match_state = match_state.inner().clone();
        let player_stats_client = player_stats_client.inner().clone();

        async move {
            for steamid in steamids {
                if let Ok(stats) = player_stats_client.get_player_stats(&steamid).await {
                    match_state.add_player_stats(steamid, stats);
                }
            }
        }
    });

    Json(ApiResponse::success("GSI data received".to_string()))
}

#[get("/match")]
pub async fn get_current_match(
    match_state: &State<MatchStateManager>,
) -> Json<ApiResponse<crate::cs2::MatchState>> {
    let state = match_state.get_state();

    if !state.is_active {
        return Json(ApiResponse::error("No active match".to_string()));
    }

    Json(ApiResponse::success(state))
}

#[post("/match/clear")]
pub async fn clear_match(
    match_state: &State<MatchStateManager>,
) -> Json<ApiResponse<String>> {
    match_state.clear();
    Json(ApiResponse::success("Match data cleared".to_string()))
}

File: server/src/controllers/cs2.ts
import { Elysia, t } from "elysia";
import {
  matchStateManager,
  playerStatsClient,
  GSIPayload,
} from "../services/cs2";

export const cs2Controller = (app: Elysia) =>
  app.group("/api/cs2", (app) =>
    app
      .post(
        "/gsi",
        async ({ body, set }) => {
          const payload = body as GSIPayload;
          const expectedToken =
            process.env.GSI_AUTH_TOKEN || "your_secret_token_here";

          if (!payload.auth || payload.auth.token !== expectedToken) {
            set.status = 401;
            return { success: false, error: "Invalid or missing auth token" };
          }

          // Only process competitive matches
          if (payload.map?.mode !== "competitive") {
            return {
              success: true,
              message: "Not a competitive match, ignoring",
            };
          }

          const mySteamId = process.env.STEAM_ID || "";
          matchStateManager.updateFromGSI(payload, mySteamId);

          // Fetch player stats asynchronously
          if (payload.allplayers) {
            const steamIds = Object.keys(payload.allplayers);

            // We don't await this to keep GSI response fast
            (async () => {
              for (const steamId of steamIds) {
                try {
                  const stats = await playerStatsClient.getPlayerStats(steamId);
                  matchStateManager.addPlayerStats(steamId, stats);
                } catch (e) {
                  console.error(`Failed to fetch stats for ${steamId}:`, e);
                }
              }
            })();
          }

          return { success: true, message: "GSI data received" };
        },
        {
          body: t.Object({
            auth: t.Optional(
              t.Object({
                token: t.String(),
              }),
            ),
            provider: t.Optional(t.Any()),
            map: t.Optional(t.Any()),
            player: t.Optional(t.Any()),
            allplayers: t.Optional(t.Any()),
            round: t.Optional(t.Any()),
          }),
          detail: {
            summary: "Receive CS2 GSI data",
            tags: ["CS2"],
          },
        },
      )
      .get(
        "/match",
        () => {
          const state = matchStateManager.getState();
          if (!state.is_active) {
            return { success: false, error: "No active match" };
          }
          return { success: true, data: state };
        },
        {
          detail: {
            summary: "Get current match state",
            tags: ["CS2"],
          },
        },
      )
      .post(
        "/match/clear",
        () => {
          matchStateManager.clear();
          return { success: true, message: "Match data cleared" };
        },
        {
          detail: {
            summary: "Clear current match data",
            tags: ["CS2"],
          },
        },
      ),
  );
